Panic on graceful ingester termination

Noticed whilst testing reboot daemon. Graceful shutdown starts at `2017-05-18T12:27:29Z`, ends with `panic` ~20 minutes later:

```
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/e6628b47e01047c9:15c1b0d1667:15c1b824666:3e81ebad in chunk cache: read tcp 10.244.204.190:40610->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/216f68277729fe76:15c1a948226:15c1b833b6e:83e67bed in chunk cache: read tcp 10.244.204.190:40526->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/f69d36a7e2a64f36:15c1ad62bb1:15c1b82f9ed:aaa76645 in chunk cache: read tcp 10.244.204.190:40126->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/14e097acd4ec1a8d:15c1a9bb03c:15c1b83c712:4bae19f8 in chunk cache: read tcp 10.244.204.190:40544->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/f840156d3c43ef48:15c1b37c5d7:15c1b834dc6:8a3bd39d in chunk cache: read tcp 10.244.204.190:40598->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/a4e585b33e2ab972:15c1ae54316:15c1b824666:b82aab85 in chunk cache: read tcp 10.244.204.190:40596->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/3b3433829b20cd50:15c1ad03cc5:15c1b8373d0:3dfc3fc8 in chunk cache: read tcp 10.244.204.190:40592->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/2c7ab806ca2ad657:15c1b2f6ad5:15c1b836a7e:8a839e03 in chunk cache: read tcp 10.244.204.190:40754->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/6f2cdc54745dd4d:15c1ab514ce:15c1b834dc6:11c11646 in chunk cache: read tcp 10.244.204.190:40608->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/d7cebfe59fb11019:15c1aba5877:15c1b834dc6:3a0a79a1 in chunk cache: read tcp 10.244.204.190:40614->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 2/a08e846aed434b97:15c1b028f7e:15c1b8246d1:63d8f8d4 in chunk cache: read tcp 10.244.204.190:40590->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/8961e8e6284849e6:15c1b0e09d4:15c1b83af05:4106da30 in chunk cache: read tcp 10.244.204.190:40618->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:22Z" level=warning msg="Could not store 1/b4142c10022a5e4a:15c1b5a03ad:15c1b82fbf5:820e8b62 in chunk cache: read tcp 10.244.204.190:40756->10.244.204.170:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:24:30Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 438.338µs" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 458.886µs" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 326.013µs" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 5.469532ms" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 302.727µs" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 315.314µs" 
time="2017-05-18T12:24:31Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 404.846µs" 
time="2017-05-18T12:24:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 280.243µs" 
time="2017-05-18T12:24:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample timestamp out of order for series ALERTS{alertname="PodRestartingTooMuch", alertstate="pending", containerStatuses="0", instance="kube-api-exporter-3558593383-8hbp1", job="monitoring/kube-api-exporter", monitor="dev", name="reboot-required-27vxu", namespace="monitoring", node="ip-172-20-1-135.ec2.internal", severity="warning"}; last timestamp: 1495110269.011, incoming timestamp: 1495110264.456) 407.2µs" 
2017/05/18 12:25:13 grpc: Server.processUnaryRPC failed to write status stream error: code = Canceled desc = "context canceled"
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.617774ms" 
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 2.00823ms" 
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 1.926956ms" 
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 3.108352ms" 
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 1.335111ms" 
time="2017-05-18T12:25:32Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.304315ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 1.339464ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.109753ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 1.227185ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 912.841µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 836.232µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 983.051µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 879.234µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.535631ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 846.107µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.013157ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 867.041µs" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_avail{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 242737668096, incoming value: 242737676288) 1.257355ms" 
time="2017-05-18T12:25:33Z" level=warning msg="gRPC /cortex.Ingester/Push (sample with repeated timestamp but different value for series node_filesystem_free{device="overlay", fstype="overlay", instance="prom-node-exporter-28rta", job="monitoring/prom-node-exporter", monitor="prod", mountpoint="/mnt/root/mnt/containers/docker/overlay2/4f9edeb63ae1d677bbf075a167eaea3fa4e7785c1a1d7a33f323e3c5d207bc65/merged", node="ip-172-20-3-251.ec2.internal"}; last value: 259135717376, incoming value: 259135725568) 1.084959ms" 
2017/05/18 12:27:00 grpc: Server.processUnaryRPC failed to write status stream error: code = Canceled desc = "context canceled"
time="2017-05-18T12:27:29Z" level=info msg="=== received SIGINT/SIGTERM ===
*** exiting" 
2017/05/18 12:27:31 transport: http2Server.HandleStreams failed to read frame: read tcp 10.244.204.190:9095->10.244.229.90:33490: use of closed network connection
time="2017-05-18T12:28:03Z" level=error msg="Error getting collectors/ring: Unexpected response code: 500" source="consul_client.go:144" 
time="2017-05-18T12:28:04Z" level=error msg="Error CASing collectors/ring, trying again 548284" source="consul_client.go:187" 
time="2017-05-18T12:28:04Z" level=error msg="Error CASing collectors/ring, trying again 548290" source="consul_client.go:187" 
time="2017-05-18T12:28:04Z" level=error msg="Error CASing collectors/ring, trying again 548291" source="consul_client.go:187" 
time="2017-05-18T12:28:04Z" level=error msg="Error CASing collectors/ring, trying again 548294" source="consul_client.go:187" 
time="2017-05-18T12:28:04Z" level=error msg="Error CASing collectors/ring, trying again 548296" source="consul_client.go:187" 
time="2017-05-18T12:28:27Z" level=warning msg="Could not store 2/15c907c7a388ac63:15c1b267eb6:15c1b8246d1:e166c71b in chunk cache: read tcp 10.244.204.190:46940->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:32:40Z" level=warning msg="Could not store 1/7a6f3c75bdcf416d:15c1a1eafa5:15c1b86b79d:bd139bc2 in chunk cache: read tcp 10.244.204.190:55262->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:34:33Z" level=warning msg="Could not store 2/1b113cfa20b7834f:15c1a95d28c:15c1b86d5cc:1231fbc3 in chunk cache: read tcp 10.244.204.190:54180->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:34:33Z" level=warning msg="Could not store 2/67396f00f6a79e14:15c1a84a804:15c1b86d5cc:f6dc824c in chunk cache: memcache: connect timeout to 10.244.254.185:11211" source="chunk_store.go:147" 
time="2017-05-18T12:34:38Z" level=warning msg="Could not store 2/a24c25e6b87355fa:15c1a9647be:15c1b86d5cc:76268d39 in chunk cache: read tcp 10.244.204.190:52300->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:34:53Z" level=warning msg="Could not store 1/145875a526ff0b48:15c1a90416a:15c1b86c2dc:f3605324 in chunk cache: read tcp 10.244.204.190:57156->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:35:16Z" level=warning msg="Could not store 1/c1175fc9bd178f7:15c1aa0e2d6:15c1b86e98f:d8c74ca7 in chunk cache: read tcp 10.244.204.190:41758->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:35:45Z" level=warning msg="Could not store 1/7c484f0f2641218c:15c1a96ee75:15c1b86cc9c:19ae0280 in chunk cache: read tcp 10.244.204.190:43482->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:37:27Z" level=warning msg="Could not store 1/1fba125b0951e12c:15c1ad30b85:15c1b86b79d:50ef347c in chunk cache: read tcp 10.244.204.190:48948->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:37:27Z" level=warning msg="Could not store 1/742fbed4b0532a38:15c1ac9391f:15c1b86bcae:b063e916 in chunk cache: read tcp 10.244.204.190:60618->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:37:49Z" level=warning msg="Could not store 1/cf1b579bf701a5ee:15c1ad195dd:15c1b86dc1e:6afcad3 in chunk cache: read tcp 10.244.204.190:42570->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:38:32Z" level=warning msg="Could not store 1/366106a67d58ea8d:15c1acc8c13:15c1b866637:1fec73ed in chunk cache: read tcp 10.244.204.190:38964->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:39:24Z" level=warning msg="Could not store 1/5601d58ce868f289:15c1aaf9748:15c1b86bd68:70994224 in chunk cache: read tcp 10.244.204.190:39302->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:39:40Z" level=warning msg="Could not store 1/d605a9d81c412dc9:15c1ad41e55:15c1b86e00e:34c612c8 in chunk cache: read tcp 10.244.204.190:52310->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:39:42Z" level=warning msg="Could not store 1/e01c5f28498d5140:15c1ad5019c:15c1b86d8f0:138217fa in chunk cache: read tcp 10.244.204.190:45130->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:40:15Z" level=warning msg="Could not store 1/2d5df6657f667429:15c1addae70:15c1b86d338:da60e578 in chunk cache: read tcp 10.244.204.190:55484->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:40:22Z" level=warning msg="Could not store 2/5f5872168dadd5fe:15c1af5c0be:15c1b86dd2e:3912988e in chunk cache: read tcp 10.244.204.190:34474->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:40:27Z" level=warning msg="Could not store 1/6a555181d2eee6a5:15c1ae2fa51:15c1b86a0cb:7fd3a5aa in chunk cache: read tcp 10.244.204.190:42060->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:40:29Z" level=warning msg="Could not store 1/f398a3fd6072f476:15c1af18db5:15c1b86c8d2:eef33eba in chunk cache: read tcp 10.244.204.190:41430->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:40:30Z" level=warning msg="Could not store 1/12d8d4e3a3c96d7:15c1add6431:15c1b86c391:e5360772 in chunk cache: memcache: connect timeout to 10.244.229.10:11211" source="chunk_store.go:147" 
time="2017-05-18T12:40:37Z" level=warning msg="Could not store 2/140f81dec43ba440:15c1ae7d2e4:15c1b86e58d:a5811278 in chunk cache: read tcp 10.244.204.190:46054->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:06Z" level=warning msg="Could not store 1/40697f15c45e1f2a:15c1af5a7bd:15c1b86c42d:3a83efae in chunk cache: read tcp 10.244.204.190:35262->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:24Z" level=warning msg="Could not store 1/5113e93f9b37e4a3:15c1afbe82e:15c1b86d696:448a88aa in chunk cache: read tcp 10.244.204.190:48272->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:34Z" level=warning msg="Could not store 1/65d3f7b224b6aa2e:15c1afa338c:15c1b86bc1c:5421ede8 in chunk cache: read tcp 10.244.204.190:47190->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:34Z" level=warning msg="Could not store 1/bc32c66283fb0ca6:15c1aef4ef3:15c1b86d402:81c2ce85 in chunk cache: read tcp 10.244.204.190:33484->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:37Z" level=warning msg="Could not store 1/a4987730f05c9f3c:15c1af248aa:15c1b86d402:4b5f24dc in chunk cache: read tcp 10.244.204.190:53180->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:41:44Z" level=warning msg="Could not store 2/25b359f11bc85b69:15c1af39434:15c1b86bffe:789ad00c in chunk cache: read tcp 10.244.204.190:59832->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:00Z" level=warning msg="Could not store 2/a1a8ffe9318e936d:15c1afa4d89:15c1b86d618:d0508c00 in chunk cache: memcache: connect timeout to 10.244.254.185:11211" source="chunk_store.go:147" 
time="2017-05-18T12:42:05Z" level=warning msg="Could not store 2/ce93d8297235e3dc:15c1af54478:15c1b86d618:1bab7176 in chunk cache: read tcp 10.244.204.190:44316->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:05Z" level=warning msg="Could not store 2/6083f55bc5ac6cba:15c1afd9a7f:15c1b86bcdb:b9e97439 in chunk cache: read tcp 10.244.204.190:49584->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:06Z" level=warning msg="Could not store 1/cc65109678855cc6:15c1b0e79fb:15c1b86de4b:651e5d17 in chunk cache: read tcp 10.244.204.190:50186->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:08Z" level=warning msg="Could not store 2/6c440493266aee2e:15c1b047862:15c1b86eec5:e0b5760a in chunk cache: read tcp 10.244.204.190:56550->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:14Z" level=warning msg="Could not store 1/2c73166b5aea3ad3:15c1af5d722:15c1b86b8fa:f0753d6c in chunk cache: read tcp 10.244.204.190:58856->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:23Z" level=warning msg="Could not store 2/19c7f448ffccdc46:15c1afd9a7f:15c1b86bcdb:4a1b1d9d in chunk cache: read tcp 10.244.204.190:33248->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:24Z" level=warning msg="Could not store 1/5c8e82efff2cbccf:15c1b008195:15c1b86dc1e:bb425894 in chunk cache: read tcp 10.244.204.190:38364->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:24Z" level=warning msg="Could not store 1/2fcb96640a0e35f0:15c1b11a64e:15c1b86d64e:dd9305ee in chunk cache: read tcp 10.244.204.190:40554->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:24Z" level=warning msg="Could not store 1/7762213d0af3968:15c1af5fe35:15c1b86e00e:c10bde33 in chunk cache: read tcp 10.244.204.190:34824->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:24Z" level=warning msg="Could not store 1/1f2f25e11167c396:15c1b09bf6d:15c1b86b79d:27255185 in chunk cache: memcache: connect timeout to 10.244.253.66:11211" source="chunk_store.go:147" 
time="2017-05-18T12:42:34Z" level=warning msg="Could not store 1/dda1e9dc9791c329:15c1b14a9c5:15c1b86e00e:e38be6b1 in chunk cache: read tcp 10.244.204.190:59392->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:34Z" level=warning msg="Could not store 1/d70ee73473eb49a2:15c1b0eb269:15c1b86dc1e:52387d73 in chunk cache: read tcp 10.244.204.190:49656->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:44Z" level=warning msg="Could not store 1/834c6276c28679fe:15c1b0eecfd:15c1b86dc1e:af8607a0 in chunk cache: read tcp 10.244.204.190:41932->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:45Z" level=warning msg="Could not store 2/db0350dbf0affb4d:15c1b1ab7c4:15c1b86c004:26c2c04b in chunk cache: read tcp 10.244.204.190:59072->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:48Z" level=warning msg="Could not store 1/c66bea3bd97dea27:15c1b0f3065:15c1b86e4ee:de7e84a5 in chunk cache: read tcp 10.244.204.190:46750->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:42:57Z" level=warning msg="Could not store 1/fddec50d6ed263b2:15c1b0c7197:15c1b86e53f:d6469b35 in chunk cache: memcache: connect timeout to 10.244.229.10:11211" source="chunk_store.go:147" 
2017/05/18 12:42:58 transport: http2Server.HandleStreams failed to read frame: read tcp 10.244.204.190:9095->10.244.253.117:50182: read: no route to host
time="2017-05-18T12:42:58Z" level=info msg="Changing ingester state from ACTIVE -> LEAVING" source="ingester_lifecycle.go:315" 
time="2017-05-18T12:42:58Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:43:00Z" level=warning msg="Could not store 1/7b5aa135f8af2480:15c1b164c75:15c1b86e895:40e305c in chunk cache: memcache: connect timeout to 10.244.254.185:11211" source="chunk_store.go:147" 
time="2017-05-18T12:43:03Z" level=warning msg="Could not store 1/93178261d17df0fd:15c1b2176a5:15c1b86a115:4225aa88 in chunk cache: read tcp 10.244.204.190:44186->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:03Z" level=warning msg="Could not store 1/6253477f47fc9dd1:15c1b096d62:15c1b86dac2:e887b542 in chunk cache: read tcp 10.244.204.190:59288->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:03Z" level=warning msg="Could not store 1/2732d80a7ac4451e:15c1b1f11a3:15c1b86c09b:bab61964 in chunk cache: memcache: connect timeout to 10.244.229.10:11211" source="chunk_store.go:147" 
time="2017-05-18T12:43:07Z" level=warning msg="Could not store 1/78fb0410e56416b1:15c1b0f7845:15c1b86b79d:b17fe4c1 in chunk cache: read tcp 10.244.204.190:48618->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:10Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:43:14Z" level=warning msg="Could not store 1/853725eb9c740168:15c1b1e082f:15c1b86a186:1315d846 in chunk cache: read tcp 10.244.204.190:51660->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:22Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:43:24Z" level=warning msg="Could not store 1/93971140aab5b444:15c1b20992e:15c1b86e895:3048c8b6 in chunk cache: read tcp 10.244.204.190:42214->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:24Z" level=warning msg="Could not store 1/f22b11b90ac14219:15c1b129a6d:15c1b86e00e:24820f9e in chunk cache: read tcp 10.244.204.190:41820->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:24Z" level=warning msg="Could not store 1/a7656b510329e470:15c1b1d4523:15c1b86c8d2:814a2f0d in chunk cache: read tcp 10.244.204.190:34014->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:28Z" level=warning msg="Could not store 1/aa11c301941f6107:15c1b1561fe:15c1b86e87e:9c662ac in chunk cache: read tcp 10.244.204.190:39546->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:30Z" level=warning msg="Could not store 2/c1b56c005b5345c2:15c1b197ed9:15c1b86e6ab:6924285 in chunk cache: memcache: connect timeout to 10.244.254.185:11211" source="chunk_store.go:147" 
time="2017-05-18T12:43:32Z" level=warning msg="Could not store 1/c3542cb19b068d0b:15c1b154d82:15c1b86d402:e7d814cc in chunk cache: read tcp 10.244.204.190:58664->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:34Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:43:35Z" level=warning msg="Could not store 1/5e1ec92b89e0835f:15c1b20560d:15c1b86e00e:db31c77e in chunk cache: read tcp 10.244.204.190:44158->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:36Z" level=warning msg="Could not store 2/6ba9fb68792d22b0:15c1b2383b6:15c1b86d96b:7c12ddb6 in chunk cache: read tcp 10.244.204.190:33902->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:41Z" level=warning msg="Could not store 2/254b54efd858020a:15c1b25047c:15c1b86c004:280cbe84 in chunk cache: read tcp 10.244.204.190:58096->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:46Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:43:53Z" level=warning msg="Could not store 1/592bc7cd4ac47912:15c1b2aef85:15c1b86b79d:c2abacf1 in chunk cache: read tcp 10.244.204.190:35756->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:55Z" level=warning msg="Could not store 2/ac6ee890996b374b:15c1b1e10f9:15c1b86e4e9:38de6cec in chunk cache: read tcp 10.244.204.190:37622->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:43:58Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:03Z" level=warning msg="Could not store 1/81a7824d937244fd:15c1b2c4575:15c1b86e895:8e923d9a in chunk cache: read tcp 10.244.204.190:44416->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:03Z" level=warning msg="Could not store 2/d2507de5bbbe9284:15c1b249e19:15c1b86ced1:9cea092 in chunk cache: read tcp 10.244.204.190:44240->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:10Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:13Z" level=warning msg="Could not store 2/3bc995ed5566e95e:15c1b2b52d4:15c1b86e055:920b4fdb in chunk cache: read tcp 10.244.204.190:37148->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:22Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:25Z" level=warning msg="Could not store 1/bcf3ba9043963eee:15c1b316959:15c1b86c8d2:9f2ba9c9 in chunk cache: read tcp 10.244.204.190:37328->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:34Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:41Z" level=warning msg="Could not store 1/8a02ddd81efc4897:15c1b38a16e:15c1b86e87e:586732d8 in chunk cache: read tcp 10.244.204.190:47854->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:45Z" level=warning msg="Could not store 2/5c867925235745da:15c1b31409e:15c1b86daaf:b789d63c in chunk cache: read tcp 10.244.204.190:48726->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:44:46Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:58Z" level=error msg="Error looking for pending ingester: no pending ingesters" source="ingester_lifecycle.go:413" 
time="2017-05-18T12:44:58Z" level=error msg="Failed to transfer chunks to another ingester: cannot find ingester to transfer chunks to: no pending ingesters" source="ingester_lifecycle.go:324" 
time="2017-05-18T12:44:59Z" level=info msg="Ingester removed from consul" source="ingester_lifecycle.go:216" 
time="2017-05-18T12:44:59Z" level=info msg="Ingester.loop() exited gracefully" source="ingester_lifecycle.go:132" 
time="2017-05-18T12:45:03Z" level=warning msg="Could not store 2/c9e2c56a7c5712a3:15c1b3b133c:15c1b86d5cc:14e29b49 in chunk cache: read tcp 10.244.204.190:44188->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:07Z" level=warning msg="Could not store 1/c9eb4065c3c4d769:15c1b3c6d51:15c1b8695b1:a4f439c3 in chunk cache: read tcp 10.244.204.190:49144->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:11Z" level=warning msg="Could not store 2/f77fead0d2bcbd15:15c1b364866:15c1b86d96b:33b44aaa in chunk cache: read tcp 10.244.204.190:48626->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:11Z" level=warning msg="Could not store 2/afb5c6509908afed:15c1b34aa9c:15c1b86d5cc:189a98af in chunk cache: read tcp 10.244.204.190:49774->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:19Z" level=warning msg="Could not store 2/1e70fa21a2ed675:15c1b3614bd:15c1b86a5bc:d0422c15 in chunk cache: read tcp 10.244.204.190:50338->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:20Z" level=warning msg="Could not store 1/95a5e258f96f93c0:15c1b4002a4:15c1b86bc1c:4b893390 in chunk cache: read tcp 10.244.204.190:50580->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:20Z" level=warning msg="Could not store 1/7f9d8e65a7cb735f:15c1b3daa7e:15c1b86e87e:952d09 in chunk cache: read tcp 10.244.204.190:35732->10.244.254.185:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:39Z" level=warning msg="Could not store 2/5b4d53a65fe17b20:15c1b43c620:15c1b86d618:dbba63e1 in chunk cache: read tcp 10.244.204.190:36578->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:39Z" level=warning msg="Could not store 2/9627a091e07d4c40:15c1b43b016:15c1b86c004:12ba5d71 in chunk cache: read tcp 10.244.204.190:47232->10.244.253.66:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:39Z" level=warning msg="Could not store 2/fdaa8741e20a17a4:15c1b42db6c:15c1b86d5cc:c9904d0b in chunk cache: read tcp 10.244.204.190:39744->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:41Z" level=warning msg="Could not store 2/51638836d7e9fbc0:15c1b437a99:15c1b86c529:b2be59d8 in chunk cache: read tcp 10.244.204.190:45160->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:45:47Z" level=warning msg="Could not store 2/b75a72ee2a46dc02:15c1b4394e6:15c1b86df76:e4fbe3bf in chunk cache: read tcp 10.244.204.190:46500->10.244.229.10:11211: i/o timeout" source="chunk_store.go:147" 
time="2017-05-18T12:46:07Z" level=error msg="Failed to flush user 1: RequestError: send request failed
caused by: Put https://weaveworks-dev-chunks.s3.amazonaws.com/1/f0df458293ade0e7%3A15c1b42ff34%3A15c1b86847c%3Aa9ed8dcc: EOF" source="ingester_flush.go:113" 
panic: enqueue on closed queue

goroutine 99 [running]:
github.com/weaveworks/cortex/pkg/util.(*PriorityQueue).Enqueue(0xc4202ad840, 0x13f4800, 0xc42b2ebf80)
	/go/src/github.com/weaveworks/cortex/pkg/util/priority_queue.go:76 +0x1cc
github.com/weaveworks/cortex/pkg/ingester.(*Ingester).flushLoop(0xc4202a2a00, 0x7)
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester_flush.go:120 +0x184
created by github.com/weaveworks/cortex/pkg/ingester.New
	/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester.go:258 +0xc91
```

This resurfaced while deleting an ingester (decreased replicas by 1):

```
time="2018-01-19T10:10:22Z" level=info msg="=== received SIGINT/SIGTERM ===\n*** exiting"
ts=2018-01-19T10:10:22.988580254Z caller=log.go:108 level=info msg="changing ingester state from" old_state=ACTIVE new_state=LEAVING
ts=2018-01-19T10:10:23.007280761Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:10:35.01122375Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:10:47.014341645Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:10:59.018711096Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:11:11.022768703Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:11:23.026234323Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:11:35.033028786Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:11:47.036756404Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:11:59.040608517Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:12:11.043938353Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:12:23.048989351Z caller=log.go:108 level=error msg="Error looking for pending ingester: %v"
ts=2018-01-19T10:12:23.049071352Z caller=log.go:108 level=error msg="Failed to transfer chunks to another ingester: %v"
ts=2018-01-19T10:13:32.626864752Z caller=log.go:108 level=error org_id=1 msg="failed to flush user" err="failed to write chunk, 1 values remaining: context deadline exceeded"
panic: enqueue on closed queue
goroutine 92 [running]:
github.com/weaveworks/cortex/pkg/util.(*PriorityQueue).Enqueue(0xc42006b600, 0x15cafe0, 0xc42e82bc20)
/go/src/github.com/weaveworks/cortex/pkg/util/priority_queue.go:86 +0x1a5
github.com/weaveworks/cortex/pkg/ingester.(*Ingester).flushLoop(0xc420436000, 0x4)
/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester_flush.go:129 +0x15a
created by github.com/weaveworks/cortex/pkg/ingester.New
/go/src/github.com/weaveworks/cortex/pkg/ingester/ingester.go:285 +0xbb7
```

--- a/pkg/ingester/ingester_lifecycle.go
+++ b/pkg/ingester/ingester_lifecycle.go
@@ -342,7 +342,7 @@ func (i *Ingester) processShutdown() {
 
 		// Close & empty all the flush queues, to unblock waiting workers.
 		for _, flushQueue := range i.flushQueues {
-			flushQueue.DrainAndClose()
+			flushQueue.DiscardAndClose()
 		}
 	}
 

--- a/pkg/util/priority_queue.go
+++ b/pkg/util/priority_queue.go
@@ -66,8 +66,8 @@ func (pq *PriorityQueue) Close() {
 	pq.cond.Broadcast()
 }
 
-// DrainAndClose closed the queue and removes all the items from it.
-func (pq *PriorityQueue) DrainAndClose() {
+// DiscardAndClose closes the queue and removes all the items from it.
+func (pq *PriorityQueue) DiscardAndClose() {
 	pq.lock.Lock()
 	defer pq.lock.Unlock()
 	pq.closed = true

--- a/pkg/util/priority_queue.go
+++ b/pkg/util/priority_queue.go
@@ -7,11 +7,12 @@ import (
 
 // PriorityQueue is a priority queue.
 type PriorityQueue struct {
-	lock   sync.Mutex
-	cond   *sync.Cond
-	closed bool
-	hit    map[string]struct{}
-	queue  queue
+	lock    sync.Mutex
+	cond    *sync.Cond
+	closing bool
+	closed  bool
+	hit     map[string]struct{}
+	queue   queue
 }
 
 // Op is an operation on the priority queue.
@@ -57,12 +58,12 @@ func (pq *PriorityQueue) Length() int {
 	return len(pq.queue)
 }
 
-// Close signals that the queue is closed. A closed queue will not accept new
-// items.
+// Close signals that the queue should be closed when it is empty.
+// A closed queue will not accept new items.
 func (pq *PriorityQueue) Close() {
 	pq.lock.Lock()
 	defer pq.lock.Unlock()
-	pq.closed = true
+	pq.closing = true
 	pq.cond.Broadcast()
 }
 
@@ -103,11 +104,12 @@ func (pq *PriorityQueue) Dequeue() Op {
 	pq.lock.Lock()
 	defer pq.lock.Unlock()
 
-	for len(pq.queue) == 0 && !pq.closed {
+	for len(pq.queue) == 0 && !(pq.closing || pq.closed) {
 		pq.cond.Wait()
 	}
 
-	if len(pq.queue) == 0 && pq.closed {
+	if len(pq.queue) == 0 && (pq.closing || pq.closed) {
+		pq.closed = true
 		return nil
 	}
 

